contracts/DomainRegistar.sol

```diff
@@ -93,60 +100,58 @@ contract DomainRegistar is Initializable {
+     * @param newPrice new domain price
+     * @param domainFullpath parent domain name starting from the top-level domain
      */
-    function registerDomain(string calldata domain) public payable {
+    function updateSubdomainPrice(uint newPrice, string calldata domainFullpath) external {
         MainStorage storage $ = _getMainStorage();
-
-        if(msg.value < $.rootEntry.weiDomainPrice) {
-            revert NotEnoughFunds(msg.value, $.rootEntry.weiDomainPrice);
-        }
-        if(!_isTopLevelDomain(domain)) {
-            revert TopLevelDomainsOnly();
-        }
-        if(!_isNewDomain(domain)) {
-            revert DuplicateDomain();
+        string[] memory domainLevels = DomainUtils.parseDomainLevels(domainFullpath);
+        DomainUtils.DomainEntry storage entry = $.rootEntry.findDomainEntry(domainLevels, domainLevels.length);
+        if(msg.sender != entry.owner) {
+            revert AccessDenied("Domain price can be changed by owner only");
         }
+        emit PriceChanged(newPrice, entry.weiDomainPrice);
+        entry.weiDomainPrice = newPrice;
+    }
 
-        DomainEntry storage entry = $.rootEntry.subdomains[domain];
-        entry.owner = payable(msg.sender);
-        entry.domainName = domain;
-        emit DomainRegistered(msg.sender, msg.sender, domain);
+    /**
+     * Register a new domain
+     * @param domainFullname domain to register provided as fullpath starting from root: mydomain.lvl1.lvl0
+     */
+    function registerDomain(string calldata domainFullname) external payable {
+        MainStorage storage $ = _getMainStorage();
+
+        string[] memory domainLevels = DomainUtils.parseDomainLevels(domainFullname);
+        string memory newSubdomainName = domainLevels[domainLevels.length - 1];
+        DomainUtils.DomainEntry storage parentEntry = $.rootEntry.findDomainEntry(domainLevels, domainLevels.length-1);
+        _validateNewDomain(parentEntry, newSubdomainName);
+        _validatePrice(parentEntry.weiDomainPrice);
+        
+        DomainUtils.DomainEntry storage newEntry = parentEntry.subdomains[newSubdomainName];
+        newEntry.owner = payable(msg.sender);
+        newEntry.domainName = newSubdomainName;
+        newEntry.weiDomainPrice = $.rootEntry.weiDomainPrice;
+        $.shares[parentEntry.owner] += parentEntry.weiDomainPrice;
+        emit DomainRegistered(msg.sender, msg.sender, domainFullname);
     }
 
     /**
      * Send all coins to the owner. Allowed for the owner only
      */
-    function withdraw() public {
+    function withdraw() external {
         MainStorage storage $ = _getMainStorage();
 
-        if(msg.sender != $.rootEntry.owner) {
-            revert AccessDenied("Only owner can request balance withdrawal");
-        }
-        $.rootEntry.owner.transfer(address(this).balance);
+        uint balance = $.shares[msg.sender];
+        $.shares[msg.sender] = 0;
+        payable(msg.sender).transfer(balance);
```

Comment:
line 147
call is better than transfer it use validation and reverts if the call fails
// - 1 point

Reply:
used `call` revering with `Error` on failure. Used error type `Error` instead of the custom one for the sake of compatibility with v1 where `transfer` throws `Error` too